### PR TITLE
Add `FeaturePropagation` as a `transform`

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
This PR implements the `FeaturePropagation` method proposed in [On the Unreasonable Effectiveness of Feature propagation in Learning on Graphs with Missing Node Features](https://arxiv.org/abs/2111.12128).

Based on the prev discussion in [this issue](https://github.com/pyg-team/pytorch_geometric/issues/4176), I implement `FeaturePropagation` as a `transform`. ~~A minimal example to reproduce the experiments is also provided.~~

### CODE REVIEW

File: torch_geometric/transforms/feature_prop.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        missing_mask (Tensor, optional): The mask indicating where
+            node features are missing. (default: :obj:`None`)
+        add_self_loops (bool, optional): If set to :obj:`False`, will not add
+            self-loops to the input graph. (default: :obj:`True`)
+        normalize (bool, optional): Whether to add self-loops and compute
+            symmetric normalization coefficients on the fly.
+            (default: :obj:`True`)
+    """
+    def __init__(self, num_iterations: int = 40,
+                 missing_mask: OptTensor = None, add_self_loops: bool = True,
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:45]: Lets make `missing_mask` a required argument.

File: torch_geometric/transforms/feature_prop.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        assert missing_mask is None or missing_mask.dtype == torch.bool
+        self.num_iterations = num_iterations
+        self.missing_mask = missing_mask
+        self.normalize = normalize
+        self.add_self_loops = add_self_loops
+
+    def __call__(self, data: Data) -> Data:
+
+        assert 'edge_index' in data or 'adj_t' in data
+
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:56]: Check if the shapes of `missing_mask` and `x` are compatible.

File: torch_geometric/transforms/feature_prop.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                edge_weight = data.edge_weight
+            edge_index = data.edge_index
+            adj_t = SparseTensor(row=edge_index[1], col=edge_index[0],
+                                 value=edge_weight,
+                                 sparse_sizes=data.size()[::-1],
+                                 is_sorted=False, trust_data=True)
+        else:
+            adj_t = data.adj_t
+
+        if self.normalize:
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:69]: I don't think normalization should be optional. According to the paper the recursion converges when the adjacency matrix is normalized.

File: torch_geometric/transforms/feature_prop.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            (default: :obj:`True`)
+    """
+    def __init__(self, num_iterations: int = 40,
+                 missing_mask: OptTensor = None, add_self_loops: bool = True,
+                 normalize: bool = True):
+        assert missing_mask is None or missing_mask.dtype == torch.bool
+        self.num_iterations = num_iterations
+        self.missing_mask = missing_mask
+        self.normalize = normalize
+        self.add_self_loops = add_self_loops
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:51]: Lets drop `add_self_loops`.

**Review by wsad1 — COMMENTED** [review]
Really appreciate you adding this. Please add a test for this feature. Left some initial comments, will take a look again in a few days.

File: torch_geometric/transforms/feature_prop.py (hunk 2 above)
Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py:56]: Thanks, will add it soon.

File: torch_geometric/transforms/feature_prop.py (hunk 3 above)
Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py:69]: Thanks for your suggestion. Will make normalization a default behavior.

File: torch_geometric/transforms/feature_prop.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        x = data.x
+        known_feature_mask = None
+        missing_mask = self.missing_mask
+        if missing_mask is not None:
+            out = torch.zeros_like(x)
+            known_feature_mask = ~missing_mask
+            out[known_feature_mask] = x[known_feature_mask]
+        else:
+            out = x.clone()
```
Comment (Padarn) [inline torch_geometric/transforms/feature_prop.py:73]: Sorry naive question: Why do we need a clone here?

File: torch_geometric/transforms/feature_prop.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        for _ in range(self.num_iterations):
+            # Diffuse current features
+            out = matmul(adj_t, out)
```
Comment (Padarn) [inline torch_geometric/transforms/feature_prop.py:77]: could we avoid defining an out above by using
```python
if self.missing_mask is not None:
   x[self.missing_mask] = 0

for _ in range(self.num_iterations):
    out = matmul(adj_t, x)
    if known_feature_mask is not None:
        # Reset original known features
        out[known_feature_mask] = x[known_feature_mask]
    x = out
```

maybe its not better

Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py:77]: Looks good to me. Updated in the latest commit, thanks!

File: torch_geometric/transforms/feature_prop.py (hunk 1 above)
Comment (emalgorithm) [reply torch_geometric/transforms/feature_prop.py]: I think the current behaviour with a `missing_mask` is counterintuitive. According to the algorithm, if no features are missing, then everything is left unchanged. On the other hand, if no features are missing (which is the case if `missing_mask` is None), then all features are diffused and change. 

I think it's easier to make `missing_mask` a required argument, so that there is only one behaviour of this algorithm.

File: torch_geometric/transforms/feature_prop.py (hunk 3 above)
Comment (emalgorithm) [reply torch_geometric/transforms/feature_prop.py]: Normalization to a symmetrically-normalized adjacency should probably happen inside the method, so that it is possible to call this method with a raw edge index, what do you think?

**Comment by emalgorithm** [comment]
Hi, the author of the paper here 👋 . Thanks a lot for working on this and let me know if you have any question!

File: torch_geometric/transforms/feature_prop.py (hunk 3 above)
Comment (wsad1) [reply torch_geometric/transforms/feature_prop.py]: Yeah i agree. looks like its in the current implementation in line 63.

File: torch_geometric/transforms/feature_prop.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        For an example of using :class:`~torch_geometric.transforms.
+        FeaturePropagation`, see `examples/feature_propagation.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        feature_propagation.py>`_.
+
+    Args:
+        num_iterations (int): The number of propagations. (default: :obj:`40`)
+        missing_mask (Tensor, optional): The mask indicating where
+            node features are missing. (default: :obj:`None`)
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:37]:
```suggestion
        missing_mask (Tensor, optional): The mask indicating where
            node features are missing, of shape `[num_nodes, channels]`. (default: :obj:`None`)
```
Add shape information.

File: torch_geometric/transforms/feature_prop.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+
+    where missing node features are inferred by known features via propagation.
+
+    .. note::
+
+        For an example of using :class:`~torch_geometric.transforms.
+        FeaturePropagation`, see `examples/feature_propagation.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        feature_propagation.py>`_.
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:32]: Not a strong preference. But i don't think we need an example for this, a code snippet with the documentation will do, like done [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.RandomLinkSplit).

**Review by wsad1 — APPROVED** [review]

File: torch_geometric/transforms/feature_prop.py (hunk 1 above)
Comment (wsad1) [reply torch_geometric/transforms/feature_prop.py]: Yeah I agree, lets drop the option to set it to `None`.

Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py]: Sure,`missing_mask` will be a required argument in the next commit.

File: torch_geometric/transforms/feature_prop.py (hunk 8 above)
Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py:32]: Indeed. Thanks for your suggestion!

File: torch_geometric/transforms/feature_prop.py (hunk 6 above)
Comment (EdisonLeeeee) [reply torch_geometric/transforms/feature_prop.py:70]: Hi @Padarn,  If I understood it correctly, this line of code `x = out` would lead to a different implementation from the original paper. According to the [original implementation](https://github.com/twitter-research/feature-propagation/blob/af2733589eab4023fca67f7e71a3b46ddbbea8cd/src/feature_propagation.py#L25-L31), `x` should always be the original node feature `data.x`, instead of the intermediate result of `out`.

I checked it and kept `x` unchanged in the latest commit.

File: torch_geometric/transforms/feature_prop.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+        from torch_geometric.transforms import FeaturePropagation
+
+        missing_mask = torch.isnan(data.x)
+        transform = FeaturePropagation(missing_mask)
+        data = transform(data)
+
+    Args:
+        missing_mask (Tensor): The mask indicating where
+            node features are missing, of shape `[num_nodes, channels]`.
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:37]:
```suggestion
        missing_mask (BoolTensor): Mask matrix
                :math:`\mathbf{M} \in {\{ 0, 1 \}}^{N\times F}`, which
                indicates where node features are missing.
```

File: torch_geometric/transforms/feature_prop.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                edge_weight = data.edge_weight
+            edge_index = data.edge_index
+            adj_t = SparseTensor(row=edge_index[1], col=edge_index[0],
+                                 value=edge_weight,
+                                 sparse_sizes=data.size()[::-1],
+                                 is_sorted=False, trust_data=True)
+        else:
+            adj_t = data.adj_t
+
+        adj_t = gcn_norm(adj_t)
```
Comment (wsad1) [inline torch_geometric/transforms/feature_prop.py:62]:
```suggestion
        adj_t = gcn_norm(adj_t, add_self_loops=False)
```

**Review by wsad1 — APPROVED** [review]
Thanks a lot for adding this. left some final comments.

File: test/transforms/test_feature_propagation.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.transforms import FeaturePropagation, ToSparseTensor
+
+
+def test_feature_propagation():
+    assert str(FeaturePropagation()) == ('FeaturePropagation('
```
[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply test/transforms/test_feature_propagation.py:8]: My bad. I'll check it again.

**Comment by EdisonLeeeee** [comment]
@wsad1 Thanks for your thorough review. Would it be able to merge now?

**Review by rusty1s — APPROVED** [review]
Thank you! Made some minor modifications.
